src/main/java/io/cdap/plugin/batch/source/MongoDBBatchSource.java

```diff
@@ -34,177 +33,201 @@
     @Nullable
     @Macro
-    private String inputQuery;
+    public String inputQuery;
 
-
-    @Name(Properties.INPUT_FIELDS)
-    @Nullable
-    @Description("A projection document limiting the fields that appear in each document. " +
-      "If no projection document is provided, all fields will be read.")
+    @Name(MongoDBConstants.ON_ERROR)
+    @Description("Specifies how to handle error in record processing. Error will be thrown if failed to parse value " +
+      "according to provided schema.")
     @Macro
-    private String inputFields;
+    public String onError;
 
-    @Name(Properties.SPLITTER_CLASS)
+    @Name(MongoDBConstants.AUTH_CONNECTION_STRING)
     @Nullable
-    @Description("The name of the Splitter class to use. If left empty, the MongoDB Hadoop Connector will attempt " +
-      "to make a best guess as to what Splitter to use.")
+    @Description("Auxiliary MongoDB connection string to authenticate against when constructing splits.")
     @Macro
-    private String splitterClass;
-
-    public MongoDBConfig(String referenceName, String connectionString, String authConnectionString,
-                         String schema, String inputQuery, String inputFields, String splitterClass) {
-      super(referenceName);
-      this.connectionString = connectionString;
-      this.authConnectionString = authConnectionString;
-      this.schema = schema;
-      this.inputQuery = inputQuery;
-      this.inputFields = inputFields;
-      this.splitterClass = splitterClass;
-    }
+    public String authConnectionString;
 
-    /**
-     * @return {@link Schema} of the dataset if one was given else null
-     * @throws IllegalArgumentException if the schema is null or not a valid JSON
-     */
     public Schema getSchema() {
-      if (schema == null) {
-        throw new IllegalArgumentException("Schema cannot be null.");
-      }
       try {
         return Schema.parseJson(schema);
       } catch (IOException e) {
-        throw new IllegalArgumentException(String.format("Unable to parse schema '%s'. Reason: %s",
-                                                         schema, e.getMessage()), e);
+        throw new InvalidConfigPropertyException("Invalid schema", e, MongoDBConstants.SCHEMA);
       }
     }
-  }
 
-  /**
-   * Property names for the config.
-   */
-  public static class Properties {
-    public static final String AUTH_CONNECTION_STRING = "authConnectionString";
-    public static final String CONNECTION_STRING = "connectionString";
-    public static final String SCHEMA = "schema";
-    public static final String INPUT_QUERY = "inputQuery";
-    public static final String INPUT_FIELDS = "inputFields";
-    public static final String SPLITTER_CLASS = "splitterClass";
+    /**
+     * Constructs JSON projection document based on the output schema.
+     * @return
+     */
+    public String getProjectionDocument() {
+      StringBuilder builder = new StringBuilder("{");
```

Comment:
Manually building a json string is brittle and not very readable. Does the mongo library have some class that represents the projection document? 

If not, it's better to create a Map<String, Integer> then use Gson to convert it to a json string.

Reply:
Changed to be:
```
    public String getProjectionDocument() {
      List<String> fieldNames = getParsedSchema().getFields().stream()
        .map(Schema.Field::getName)
        .collect(Collectors.toList());
	
      return Projections.include(fieldNames).toString();
    }
```

Thanks.